genume/registry/registry.py

```diff
@@ -1,47 +1,150 @@
+        for c in children:
+            c.finish_up()
+
+    def run(self):
+        log.info("Registry thread is starting up...")
+        while True:
+            if self.lock.acquire(blocking=True) and self.state is RegistryStates.start:
+                log.debug("A refresh has been requested.")
+                # Actual refresh starts here.
+                # First scan the directories.
+                self.state = RegistryStates.scanning
+                pending_children = self.scan_ahead()
+                # Take a break to process events.
+                GObject.idle_add(make_signal_emitter(self.observer, "scan_complete"))
+                # Start collecting and parsing commands in steps,
+                # so the main thread is not getting blocked.
+                self.state = RegistryStates.collecting
+                self.execute_children(pending_children)
+                # Change state and prepare to hand data to main thread.
+                self.state = RegistryStates.finish
+                # Inform main thread.
+                GObject.idle_add(make_signal_emitter(self.observer, "refresh_complete"))
+                log.debug("Registry refresh has been completed.")
+
+    def request_refresh(self):
+        "This method reloads the enumeration by running all the scripts(async version)."
+        if self.state is RegistryStates.waiting:
+            self.state = RegistryStates.start
+            self.lock.release()
+            return True
+        else:
+            raise RuntimeError("Refresh requested but registry is in an invalid state!")
+
+    def get_async_data(self):
+        "Gets data(an Entry tree) after a refresh request has finished. Else returns false."
+        if self.state is RegistryStates.finish:
+            # Reset state to prepare for next refresh.
+            self.state = RegistryStates.waiting
+            return self.root
+        else:
+            return False
 
-    def __init__(self):
-        # Initialize root.
-        self.root = CategoryEntry()
-
-    def update(self):
-        "This method reloads the enumeration by running all the scripts."
-        self.root = CategoryEntry()
-        self._handle_dir(self.root, Path(SCRIPTS_ROOT))
-
-    @staticmethod
-    def _handle_dir(cat, path):
-        "Internal static recursive method to find scripts."
-        for file in sorted(path.iterdir()):
-            if match_list(file.name, SCRIPTS_IGNORE):
-                # Ignore case.
-                pass
-            elif file.is_dir():
-                # Directory case.
-                new_cat = CategoryEntry(cat)
-                key = file.name
-                cat[key] = new_cat
-                Registry._handle_dir(new_cat, file)
-            elif file.suffix == ".py":
-                # Python script case.
-                PythonScript.runscript(cat, file)
-            else:
-                # Generic executable case.
-                run_and_parse(cat, file)
+    def refresh(self):
```

Comment:
There are also still some calls to the old `update()` instead of the new `refresh()`.

Reply:
Good luck!!!
Hint: you cannot use the 'easy' refresh on the gui because a glib main loop is already running (Gtk.main).